virtual_rainforest/core/axes.py

```diff
@@ -1,323 +1,384 @@
+class AxisValidator(ABC):
+    """The AxisValidator abstract base class.
 
-Users can register their own functions to map data onto a core axis using the
-:func:`~virtual_rainforest.core.data.register_axis_validator` decorator. The
-function itself should have the following signature:
+    This ABC provides the structure for axis validators. These are used to check that a
+    DataArray to be added to a Data instance is congruent with the configuration of a
+    virtual rainforest simulation. The base class provides abstract methods that provide
+    the following functionality:
 
-    func(data: Data, darray: DataArray) -> DataArray
-"""
+    * Test if the validator subclass can be applied to a particular DataArray
+      (:meth:`~virtual_rainforest.core.axes.AxisValidator.can_validate`).
+    * Run any appropriate validation on DataArrays that pass the validation test
+      (:meth:`~virtual_rainforest.core.axes.AxisValidator.run_validation`).
 
+    The class method (:meth:`~virtual_rainforest.core.axes.AxisValidator.validate`)
+    wraps the subclass specific implementations of these two functions. If the input can
+    be validated then it returns the DataArray with any validation applied, otherwise
+    the original input is returned. The method also sets the attributes of validated
+    DataArrays to record that validation has been passed on the core axis.
+    """
 
-def register_axis_validator(axis: str, signature: tuple) -> Callable:
-    """Add a new validator and data signature to a core axis.
+    core_axis: str = ""
+    dim_names: set[str] = {""}
 
-    This decorator adds the decorated method to the AXIS_VALIDATORS registry to extend
-    or amend the available routines used to load and validate an array containing
-    spatial information. The decorator adds the method to the list of validators
-    provided for the given axis class attribute under the given signature.
+    def __init__(self) -> None:
+        super().__init__()
 
-    The signature consists of a tuple containing:
+    @classmethod
+    def __init_subclass__(cls) -> None:
+        """Adds new subclasses to the AxisValidator registry.
 
-    * a tuple of dimension names required in the data array to use the function,
-    * a tuple of coordinate names required in the data array to use the function - this
-      can be an empty tuple if no coordinate names are required,
-    * a tuple of grid types that the loader can be used with. The "any" grid type
-      can be used to indicate that a loader will work with any grid type.
+        When new subclasses are created this method automatically extends the
+        :var:`~virtual_rainforest.core.axes.AXIS_VALIDATORS` registry. The subclass is
+        added to the list of AxisValidators that apply to the subclass core axis.
+        """
 
-    Args:
-        axis: The core axis that the validator applies to
-        signature: A tuple of the validator signature.
-    """
+        if cls.core_axis == "":
```

Comment:
Shouldn't you also check that the `dim_names` are not an empty set?

Reply:
That's fine, let's not overthink tests! Parametric testing is useful when parameters are simple stuff, otherwise it is not worth the effort and, indeed, masks what you are actually testing.